### PR TITLE
Send frame diffs over instead of full frames

### PR DESCRIPTION
**Pull request by ebetica**
Now the connection also uses frame diffs instead of the full frames.

I ran clang-format on every file this change touched, so a lot of this is automatic, sorry guys :( I'll split it into 2 commits next time. module.cc wasn't really changed much except for the destructor. 

In frame.h I removed all of the diff functions involving references, so we don't get an explosion of signatures that take either pointers or references,

### CODE REVIEW

File: client/state.cpp [hunk 1]
```diff
@@ -206,13 +206,28 @@ std::vector<std::string> State::update(
   return upd;
 }
 
+void update_frame(State* state, const torchcraft::fbs::FrameData* fd) {
```
Comment (jgehring) [inline client/state.cpp]: Looks like this should be a member function?

File: include/frame.h [hunk 2]
```diff
@@ -347,12 +347,12 @@ class FrameDiff {
...
-Frame* frame_undiff(FrameDiff& lhs, Frame& rhs);
-Frame* frame_undiff(Frame& lhs, FrameDiff& rhs);
-Frame* frame_undiff(FrameDiff* lhs, Frame* rhs);
-Frame* frame_undiff(Frame* lhs, FrameDiff* rhs);
+FrameDiff frame_diff(Frame&, Frame&);
+FrameDiff frame_diff(Frame*, Frame*);
+Frame* frame_undiff(FrameDiff*, Frame*);
+Frame* frame_undiff(Frame*, FrameDiff*);
+void frame_undiff(Frame*, FrameDiff*, Frame*);
+void frame_undiff(Frame*, Frame*, FrameDiff*);
```
Comment (jgehring) [inline include/frame.h]: It would be nice to provide `from` and `to` argument names here to emphasize the direction of the diff.

Comment (ebetica) [inline include/frame.h]: Undiffs (addition) is commutative :)

**Comment by ebetica** [comment]
Timing until [this line](https://github.com/TorchCraft/TorchCraft/blob/develop/examples/dump_replay.lua#L55) on bwrep_gwhxe.rep for the first 5000 frames:

With diffs:
th dump_replay.lua -t localhost  11.05s user 0.58s system 44% cpu 26.142 total

With diffs but without reserializing the frame string:
th dump_replay.lua -t localhost  3.47s user 0.54s system 21% cpu 18.475 total

Without diffs:
th dump_replay.lua -t localhost  10.80s user 0.63s system 35% cpu 32.410 total

What do we need the [frame string](https://github.com/TorchCraft/TorchCraft/pull/167/files#diff-4c76f9fac82dd94513c6286362f89c33R216) for and can I remove it @syhw @jgehring

**Pull request merged**